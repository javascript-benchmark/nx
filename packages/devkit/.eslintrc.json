{
  "extends": "../../.eslintrc",
  "ignorePatterns": ["!**/*"],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "rules": {
        "@typescript-eslint/explicit-module-boundary-types": ["error"],
        "@typescript-eslint/no-restricted-imports": [
          "error",
          {
            "paths": [
              "@nx/workspace",
              "@angular-devkit/core",
              "@angular-devkit/architect",
              "@angular-devkit/schematics"
            ],
            "patterns": [
              {
                "group": ["nx/**/*"],
                "message": "Use requireNx() from packages/devkit/nx.ts OR use a type import instead.",
                "allowTypeImports": true
              },
              {
                "group": ["@nx/devkit/**/*"],
                "message": "Use a relative import"
              }
            ]
          }
        ]
      }
    },
    {
      "files": ["*.spec.ts"],
      "rules": {
        "@typescript-eslint/no-restricted-imports": "off"
      }
    },
    {
      "files": ["*.ts", "*.tsx"],
      "rules": {}
    },
    {
      "files": ["*.js", "*.jsx"],
      "rules": {}
    },
    {
      "files": ["./package.json"],
      "parser": "jsonc-eslint-parser",
      "rules": {
        "@nx/nx-plugin-checks": "error"
      }
    }
  ]
}
