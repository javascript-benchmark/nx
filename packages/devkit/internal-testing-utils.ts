/* eslint-disable @typescript-eslint/no-restricted-imports */
export * from 'nx/internal-testing-utils/assert-valid-migrations';
export * from 'nx/internal-testing-utils/run-migration-against-this-workspace';
export * from 'nx/internal-testing-utils/with-environment';
