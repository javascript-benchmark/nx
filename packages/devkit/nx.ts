// After Nx v18, this can be removed and replaced with either:
// - import {} from 'nx/src/devkit-exports'
// - import {} from 'nx/src/devkit-internals'
export function requireNx(): typeof import('nx/src/devkit-exports') &
  Partial<typeof import('nx/src/devkit-internals')> {
  try {
    let result = { ...require('nx/src/devkit-exports') };
    try {
      result = {
        ...result,
        // Remove in Nx v18, devkit should not support Nx v16.0.2 at that point.
        ...require('nx/src/devkit-internals'),
      };
    } catch {}
    return result;
  } catch {
    // Remove in Nx V17, devkit should not support Nx < 16 at that point.
    return require('./nx-reexports-pre16');
  }
}
