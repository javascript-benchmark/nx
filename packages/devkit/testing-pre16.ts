/* eslint-disable @typescript-eslint/no-restricted-imports */
export {
  createTreeWithEmptyWorkspace,
  createTreeWithEmptyV1Workspace,
} from 'nx/src/generators/testing-utils/create-tree-with-empty-workspace';
export { createTree } from 'nx/src/generators/testing-utils/create-tree';
