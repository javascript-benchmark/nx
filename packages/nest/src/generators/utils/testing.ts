import { addProjectConfiguration, Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';

export function createTreeWithNestApplication(appName: string): Tree {
  const tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  addProjectConfiguration(tree, appName, {
    root: `apps/${appName}`,
    sourceRoot: `apps/${appName}/src`,
    projectType: 'application',
    targets: {},
  });
  return tree;
}
