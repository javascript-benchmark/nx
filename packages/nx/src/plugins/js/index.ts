import {
  ProjectGraph,
  ProjectGraphProcessor,
} from '../../config/project-graph';
import { ProjectGraphBuilder } from '../../project-graph/project-graph-builder';
import { buildExplicitDependencies } from './project-graph/build-dependencies/build-dependencies';
import { readNxJson } from '../../config/configuration';
import { fileExists, readJsonFile } from '../../utils/fileutils';
import { PackageJson } from '../../utils/package-json';
import {
  lockFileExists,
  lockFileHash,
  parseLockFile,
} from './lock-file/lock-file';
import { NrwlJsPluginConfig, NxJsonConfiguration } from '../../config/nx-json';
import { dirname, join } from 'path';
import { projectGraphCacheDirectory } from '../../utils/cache-directory';
import { readFileSync, writeFileSync } from 'fs';
import { workspaceRoot } from '../../utils/workspace-root';
import { ensureDirSync } from 'fs-extra';

export const processProjectGraph: ProjectGraphProcessor = async (
  graph,
  context
) => {
  const builder = new ProjectGraphBuilder(graph, context.fileMap);
  const pluginConfig = jsPluginConfig(readNxJson());

  if (pluginConfig.analyzePackageJson) {
    // during the create-nx-workspace lock file might not exists yet
    if (lockFileExists()) {
      const lockHash = lockFileHash();
      let parsedLockFile: ProjectGraph;
      if (lockFileNeedsReprocessing(lockHash)) {
        parsedLockFile = parseLockFile();
        writeLastProcessedLockfileHash(lockHash, parsedLockFile);
      } else {
        parsedLockFile = readParsedLockFile();
      }
      builder.mergeProjectGraph(parsedLockFile);
    }
  }

  await buildExplicitDependencies(pluginConfig, context, builder);

  return builder.getUpdatedProjectGraph();
};

const lockFileHashFile = join(projectGraphCacheDirectory, 'lockfile.hash');
const parsedLockFile = join(
  projectGraphCacheDirectory,
  'parsed-lock-file.json'
);

function lockFileNeedsReprocessing(lockHash: string) {
  try {
    return readFileSync(lockFileHashFile).toString() !== lockHash;
  } catch {
    return true;
  }
}

function writeLastProcessedLockfileHash(hash: string, lockFile: ProjectGraph) {
  ensureDirSync(dirname(lockFileHashFile));
  writeFileSync(parsedLockFile, JSON.stringify(lockFile, null, 2));
  writeFileSync(lockFileHashFile, hash);
}

function readParsedLockFile(): ProjectGraph {
  return JSON.parse(readFileSync(parsedLockFile).toString());
}

function jsPluginConfig(
  nxJson: NxJsonConfiguration
): Required<NrwlJsPluginConfig> {
  const nxJsonConfig: NrwlJsPluginConfig =
    nxJson?.pluginsConfig?.['@nx/js'] ?? nxJson?.pluginsConfig?.['@nrwl/js'];

  if (nxJsonConfig) {
    return {
      analyzePackageJson: true,
      analyzeSourceFiles: true,
      ...nxJsonConfig,
    };
  }

  if (!fileExists(join(workspaceRoot, 'package.json'))) {
    return {
      analyzePackageJson: false,
      analyzeSourceFiles: false,
    };
  }

  const packageJson = readJsonFile<PackageJson>(
    join(workspaceRoot, 'package.json')
  );

  const packageJsonDeps = {
    ...packageJson.dependencies,
    ...packageJson.devDependencies,
  };
  if (
    packageJsonDeps['@nx/workspace'] ||
    packageJsonDeps['@nx/js'] ||
    packageJsonDeps['@nx/node'] ||
    packageJsonDeps['@nx/next'] ||
    packageJsonDeps['@nx/react'] ||
    packageJsonDeps['@nx/angular'] ||
    packageJsonDeps['@nx/web'] ||
    packageJsonDeps['@nrwl/workspace'] ||
    packageJsonDeps['@nrwl/js'] ||
    packageJsonDeps['@nrwl/node'] ||
    packageJsonDeps['@nrwl/next'] ||
    packageJsonDeps['@nrwl/react'] ||
    packageJsonDeps['@nrwl/angular'] ||
    packageJsonDeps['@nrwl/web']
  ) {
    return { analyzePackageJson: true, analyzeSourceFiles: true };
  } else {
    return { analyzePackageJson: true, analyzeSourceFiles: false };
  }
}
