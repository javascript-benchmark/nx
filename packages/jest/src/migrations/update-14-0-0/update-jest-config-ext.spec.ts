import {
  readJson,
  readProjectConfiguration,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { libraryGenerator as workspaceLib } from '@nx/js';
import jestInitGenerator from '../../generators/init/init';
import { updateJestConfigExt } from './update-jest-config-ext';

const setupDefaults = {
  js: true,
  skipPackageJson: true,
  libName: 'lib-one',
  setParserOptionsProject: false,
};
const oldConfig = `
module.exports = {
  displayName: 'PLACE_HOLDER',
  preset: '../../jest.preset.js',
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    }
  },
  transform: {
    '^.+\\\\.[tj]sx?$': 'ts-jest'
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/PLACE_HOLDER'
};
`;

async function libSetUp(tree: Tree, options = setupDefaults) {
  jestInitGenerator(tree, {
    js: options.js,
    skipPackageJson: options.skipPackageJson,
  });
  await workspaceLib(tree, {
    name: options.libName,
    setParserOptionsProject: options.setParserOptionsProject,
  });
  tree.rename(
    `libs/${options.libName}/jest.config.ts`,
    `libs/${options.libName}/jest.config.js`
  );
  tree.write(
    `libs/${options.libName}/jest.config.js`,
    oldConfig.replace(/PLACE_HOLDER/g, options.libName)
  );
  updateProjectConfiguration(tree, options.libName, {
    ...readProjectConfiguration(tree, options.libName),
    targets: {
      test: {
        executor: '@nrwl/jest:jest',
        options: {
          jestConfig: `libs/${options.libName}/jest.config.js`,
          passWithNoTests: true,
        },
        configurations: {
          production: {
            silent: true,
          },
        },
      },
    },
  });
}

describe('Jest Migration (v14.0.0)', () => {
  let tree: Tree;
  beforeEach(async () => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  });

  it('should rename project jest.config.js to jest.config.ts', async () => {
    await libSetUp(tree);

    await updateJestConfigExt(tree);
    expect(tree.exists('libs/lib-one/jest.config.ts')).toBeTruthy();
    expect(tree.read('libs/lib-one/jest.config.ts', 'utf-8')).toMatchSnapshot();
  });

  it('should rename root jest.config.js', async () => {
    await libSetUp(tree);

    await updateJestConfigExt(tree);
    expect(tree.exists('jest.config.ts')).toBeTruthy();
    expect(tree.exists('jest.preset.js')).toBeTruthy();
  });

  it('should NOT update jest.config.ts preset', async () => {
    await libSetUp(tree);

    tree.rename('libs/lib-one/jest.config.js', 'libs/lib-one/jest.config.ts');
    const projectConfig = readProjectConfiguration(tree, 'lib-one');
    updateProjectConfiguration(tree, 'lib-one', {
      ...projectConfig,
      targets: {
        test: {
          ...projectConfig.targets.test,
          options: {
            jestConfig: 'libs/lib-one/jest.config.ts',
            passWithNoTests: true,
          },
        },
      },
    });
    expect(tree.exists('libs/lib-one/jest.config.ts')).toBeTruthy();
    await updateJestConfigExt(tree);
    expect(tree.exists('libs/lib-one/jest.config.ts')).toBeTruthy();
    expect(tree.read('libs/lib-one/jest.config.ts', 'utf-8')).toMatchSnapshot();
  });

  it('should only update js/ts files', async () => {
    await libSetUp(tree);

    tree.rename('libs/lib-one/jest.config.js', 'libs/lib-one/jest.config.ts');
    updateProjectConfiguration(tree, 'lib-one', {
      ...readProjectConfiguration(tree, 'lib-one'),
      targets: {
        test: {
          executor: '@nrwl/jest:jest',
          options: {
            jestConfig: 'libs/lib-one/jest.config.ts',
            passWithNoTests: true,
          },
        },
      },
    });

    await libSetUp(tree, { ...setupDefaults, libName: 'lib-two' });
    tree.delete('libs/lib-two/jest.config.ts'); // lib generator creates a ts file
    tree.write('libs/lib-two/jest.config.json', '{}');
    updateProjectConfiguration(tree, 'lib-two', {
      ...readProjectConfiguration(tree, 'lib-two'),
      targets: {
        test: {
          executor: '@nrwl/jest:jest',
          options: {
            jestConfig: 'libs/lib-two/jest.config.json',
            passWithNoTests: true,
          },
        },
      },
    });

    await libSetUp(tree, { ...setupDefaults, libName: 'lib-three' });
    expect(tree.exists('libs/lib-one/jest.config.ts')).toBeTruthy();
    await updateJestConfigExt(tree);
    expect(tree.exists('libs/lib-one/jest.config.ts')).toBeTruthy();
    expect(tree.exists('libs/lib-two/jest.config.ts')).toBeFalsy();
    expect(tree.exists('libs/lib-two/jest.config.json')).toBeTruthy();
    expect(tree.exists('libs/lib-three/jest.config.ts')).toBeTruthy();
  });

  it('should not throw error if file does not exit', async () => {
    await libSetUp(tree);

    tree.delete('libs/lib-one/jest.config.js');
    await updateJestConfigExt(tree);
    expect(tree.exists('libs/lib-one/jest.config.ts')).toBeFalsy();
    expect(tree.exists('libs/lib-one/jest.config.js')).toBeFalsy();
  });

  it('should update correct tsconfigs', async () => {
    await libSetUp(tree);

    updateJson(tree, 'libs/lib-one/tsconfig.lib.json', (json) => {
      json.exclude = ['src/**/*.spec.ts'];
      return json;
    });

    updateJson(tree, 'libs/lib-one/tsconfig.spec.json', (json) => {
      json.include = ['src/**/*.spec.ts'];
      return json;
    });

    await updateJestConfigExt(tree);

    const tsconfig = readJson(tree, 'libs/lib-one/tsconfig.json');
    const libTsConfig = readJson(tree, 'libs/lib-one/tsconfig.lib.json');
    const specTsConfig = readJson(tree, 'libs/lib-one/tsconfig.spec.json');

    expect(tsconfig.exclude).toBeFalsy();
    expect(libTsConfig.exclude).toEqual(['src/**/*.spec.ts', 'jest.config.ts']);
    expect(specTsConfig.exclude).toBeFalsy();
    expect(specTsConfig.include).toEqual([
      'src/**/*.spec.ts',
      'jest.config.ts',
    ]);
  });

  it('should add exclude to root tsconfig with no references', async () => {
    await libSetUp(tree);

    tree.delete('libs/lib-one/tsconfig.spec.json');
    tree.delete('libs/lib-one/tsconfig.lib.json');

    updateJson(tree, 'libs/lib-one/tsconfig.json', (json) => {
      delete json.references;
      return json;
    });

    await updateJestConfigExt(tree);

    const tsconfig = readJson(tree, 'libs/lib-one/tsconfig.json');

    expect(tsconfig.exclude).toEqual(['jest.config.ts']);
    expect(tree.exists('libs/lib-one/tsconfig.spec.json')).toBeFalsy();
    expect(tree.exists('libs/lib-one/tsconfig.lib.json')).toBeFalsy();
  });

  it('should update the excludes of next js apps using the project parser settings', async () => {
    await libSetUp(tree, { ...setupDefaults, setParserOptionsProject: true });

    const projectConfig = readProjectConfiguration(tree, 'lib-one');
    projectConfig.targets['build'] = {
      executor: '@nrwl/next:build',
      options: {},
    };
    updateProjectConfiguration(tree, 'lib-one', projectConfig);
    updateJson(tree, 'libs/lib-one/tsconfig.json', (json) => {
      // simulate nextJS tsconfig;
      json.exclude = ['node_modules'];
      return json;
    });
    const esLintJson = readJson(tree, 'libs/lib-one/.eslintrc.json');
    // make sure the parserOptions are set correctly
    expect(esLintJson.overrides[0]).toMatchSnapshot();

    await updateJestConfigExt(tree);

    const tsconfigSpec = readJson(tree, 'libs/lib-one/tsconfig.spec.json');
    expect(tsconfigSpec.exclude).toEqual(['node_modules']);
  });

  it('should produce the same results when running multiple times', async () => {
    await libSetUp(tree);
    updateJson(tree, 'libs/lib-one/tsconfig.lib.json', (json) => {
      json.exclude = ['src/**/*.spec.ts'];
      return json;
    });
    updateJson(tree, 'libs/lib-one/tsconfig.spec.json', (json) => {
      json.include = ['src/**/*.spec.ts'];
      return json;
    });

    await setupNextProj(tree);

    const esLintJson = readJson(tree, 'libs/my-next-proj/.eslintrc.json');
    // make sure the parserOptions are set correctly for next
    expect(esLintJson.overrides[0]).toMatchSnapshot();

    await updateJestConfigExt(tree);

    assertNextProj(tree);
    assertLib(tree);

    await updateJestConfigExt(tree);

    assertNextProj(tree);
    assertLib(tree);

    expect(tree.read('libs/lib-one/jest.config.ts', 'utf-8')).toMatchSnapshot();
  });
});

async function setupNextProj(tree: Tree) {
  await libSetUp(tree, {
    ...setupDefaults,
    libName: 'my-next-proj',
    setParserOptionsProject: true,
  });
  const projectConfig = readProjectConfiguration(tree, 'my-next-proj');
  projectConfig.targets['build'] = {
    executor: '@nrwl/next:build',
    options: {},
  };
  updateProjectConfiguration(tree, 'my-next-proj', projectConfig);
  updateJson(tree, 'libs/my-next-proj/tsconfig.json', (json) => {
    // simulate nextJS tsconfig;
    json.exclude = ['node_modules'];
    return json;
  });
}

function assertNextProj(tree: Tree) {
  expect(
    readJson(tree, 'libs/my-next-proj/tsconfig.spec.json').exclude
  ).toEqual(['node_modules']);
  expect(
    readJson(tree, 'libs/my-next-proj/tsconfig.spec.json').include
  ).toEqual(expect.arrayContaining(['jest.config.ts']));
}

function assertLib(tree: Tree) {
  expect(readJson(tree, 'libs/lib-one/tsconfig.json').exclude).toBeFalsy();
  expect(readJson(tree, 'libs/lib-one/tsconfig.spec.json').exclude).toBeFalsy();
  expect(readJson(tree, 'libs/lib-one/tsconfig.spec.json').include).toEqual([
    'src/**/*.spec.ts',
    'jest.config.ts',
  ]);

  expect(readJson(tree, 'libs/lib-one/tsconfig.lib.json').exclude).toEqual([
    'src/**/*.spec.ts',
    'jest.config.ts',
  ]);
  expect(readJson(tree, 'libs/lib-one/tsconfig.lib.json').include).toEqual([
    'src/**/*.ts',
  ]);
}
