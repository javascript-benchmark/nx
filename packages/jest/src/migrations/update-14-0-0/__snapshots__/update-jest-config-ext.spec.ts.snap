// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Migration (v14.0.0) should NOT update jest.config.ts preset 1`] = `
"/* eslint-disable */

module.exports = {
  displayName: 'lib-one',
  preset: '../../jest.preset.js',
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    },
  },
  transform: {
    '^.+\\\\.[tj]sx?$': 'ts-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/lib-one',
};
"
`;

exports[`Jest Migration (v14.0.0) should produce the same results when running multiple times 1`] = `
{
  "files": [
    "*.ts",
    "*.tsx",
    "*.js",
    "*.jsx",
  ],
  "parserOptions": {
    "project": [
      "libs/my-next-proj/tsconfig.*?.json",
    ],
  },
  "rules": {},
}
`;

exports[`Jest Migration (v14.0.0) should produce the same results when running multiple times 2`] = `
"/* eslint-disable */
module.exports = {
  displayName: 'lib-one',
  preset: '../../jest.preset.js',
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    },
  },
  transform: {
    '^.+\\\\.[tj]sx?$': 'ts-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/lib-one',
};
"
`;

exports[`Jest Migration (v14.0.0) should rename project jest.config.js to jest.config.ts 1`] = `
"/* eslint-disable */

module.exports = {
  displayName: 'lib-one',
  preset: '../../jest.preset.js',
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    },
  },
  transform: {
    '^.+\\\\.[tj]sx?$': 'ts-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/lib-one',
};
"
`;

exports[`Jest Migration (v14.0.0) should update the excludes of next js apps using the project parser settings 1`] = `
{
  "files": [
    "*.ts",
    "*.tsx",
    "*.js",
    "*.jsx",
  ],
  "parserOptions": {
    "project": [
      "libs/lib-one/tsconfig.*?.json",
    ],
  },
  "rules": {},
}
`;
